Add `shards` to `valid_encodings` to enable sharded Zarr writing

jacobbieker opened the pull request:
Adds `shards` to the list of `valid_encodings` in the zarr backend, so that sharded Zarr V3s can be written.

- [x] Closes #9947

jhamman commented:
Thanks @jacobbieker for opening a PR. Do you have time to add a roundtrip test for this? This test would be a good test to emulate:

https://github.com/pydata/xarray/blob/6bea71572120c0eef2e4bf8d269d32b17a39dc35/xarray/tests/test_backends.py#L2486-L2494

jacobbieker commented:
> Thanks @jacobbieker for opening a PR. Do you have time to add a roundtrip test for this? This test would be a good test to emulate:
> 
> https://github.com/pydata/xarray/blob/6bea71572120c0eef2e4bf8d269d32b17a39dc35/xarray/tests/test_backends.py#L2486-L2494

Yep! I think I've added one that works for that, and updated the Zarr V3 loading to include the `shards`. It seems to pass the tests locally so far at least.

jhamman commented:
I seem to remember fixing this in the Dask test suite by manually cleaning up thread resources. This may help: https://github.com/dask/dask/blob/e04734b4d8959ba259801f2e2a490cb4ee8d891f/dask/tests/test_distributed.py#L338-L358

kmuehlbauer commented:
@jhamman The dask zarr integration test is also failing on main. Would be great to have a fix applied.

kmuehlbauer commented:
> I seem to remember fixing this in the Dask test suite by manually cleaning up thread resources. This may help: https://github.com/dask/dask/blob/e04734b4d8959ba259801f2e2a490cb4ee8d891f/dask/tests/test_distributed.py#L338-L358

See #9967

jacobbieker commented:
> > I seem to remember fixing this in the Dask test suite by manually cleaning up thread resources. This may help: https://github.com/dask/dask/blob/e04734b4d8959ba259801f2e2a490cb4ee8d891f/dask/tests/test_distributed.py#L338-L358
> 
> See #9967

@jhamman yeah, this change in #9967 fixes the tests and the failing ones pass.

jhamman reviewed (approved):
Thanks for this @jacobbieker !

jhamman commented:
We'll let @dcherian do a final review and/or merge.

dcherian commented:
Looks great. Thanks @jacobbieker, welcome to Xarray!

jhamman commented:
Thanks @jacobbieker for the contribution here!
